src/NodeServices/PythonServices.cs

```diff
@@ -0,0 +1,326 @@
+        /// </summary>
+        /// <param name="engine"></param>
+        /// <param name="evaluatorClass"></param>
+        /// <param name="evaluationMethod"></param>
+        internal void GetEvaluatorInfo(PythonEngineVersion engine, out string evaluatorClass, out string evaluationMethod)
+        {
+            // Provide evaluator info when the selected engine is loaded
+            if (engine == PythonEngineVersion.IronPython2 &&
+                AvailableEngines.Any(x => x.Version == PythonEngineVersion.IronPython2))
+            {
+                evaluatorClass = IronPythonEvaluatorClass;
+                evaluationMethod = IronPythonEvaluationMethod;
+                return;
+            }
+            if (engine == PythonEngineVersion.CPython3 &&
+                AvailableEngines.Any(x => x.Version == PythonEngineVersion.CPython3))
+            {
+                evaluatorClass = CPythonEvaluatorClass;
+                evaluationMethod = CPythonEvaluationMethod;
+                return;
+            }
+
+            // Throwing at the compilation stage is handled as a non-retryable error by the Dynamo engine.
+            // Instead, we want to produce an error at the evaluation stage, so we can eventually recover.
+            // We handle this by providing a dummy Python evaluator that will evaluate to an error message.
+            evaluatorClass = DummyEvaluatorClass;
+            evaluationMethod = DummyEvaluatorMethod;
+        }
+
+        private void AssemblyLoadEventHandler(object sender, AssemblyLoadEventArgs args)
+        {
+            PythonEngineManager.Instance.ScanPythonEngine(args.LoadedAssembly);
+        }
+
+        private void ScanPythonEngine(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                return;
+            }
+
+            if (!AvailableEngines.Any(x => x.Version == PythonEngineVersion.IronPython2))
+            {
+                var engine = CreateEngineProxy(assembly, PythonEngineVersion.IronPython2);
+                if (engine != null)
+                {
+                    AvailableEngines.Add(engine);
+                }
+            }
+
+            if (!AvailableEngines.Any(x => x.Version == PythonEngineVersion.CPython3))
+            {
+                var engine = CreateEngineProxy(assembly, PythonEngineVersion.CPython3);
+                if (engine != null)
+                {
+                    AvailableEngines.Add(engine);
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Proxy class for any Python Engine that can be loaded by Dynamo
+    /// Provides basic APIs to interact with the Python Engine
+    /// </summary>
+    [SupressImportIntoVM]
+    public class PythonEngineProxy : IDisposable
+    {
+        /// <summary>
+        /// The version of the Python Engine connected to this proxy instance
+        /// </summary>
+        public readonly PythonEngineVersion Version;
```

Comment:
I think it's advised by MSDN to use a property instead of a field in most cases for public members.